File: leetcode/77.组合/77.组合.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        track = []
        res = []

        def back_track(start):
            # 满足结束条件，路径数量多大限制数量
            # 多大树底部
            if len(track) == k:
                res.append(list(track))
                return
            # 遍历选择，范围[1~n]
            for i in range(start, n + 1):
                # 做选择
                track.append(i)
                # dfs
                back_track(i + 1)
                # 撤销选择
                track.pop()

        # 范围[1~n] 所以从1开始
        back_track(1)
        return res


if __name__ == '__main__':
    n = 4
    k = 2
    result = Solution().combine(n, k)
    print(result)

File: leetcode/77.组合/77.组合.go
package main

import "fmt"

func combine(n int, k int) [][]int {
	track := make([]int, 0, k)
	res := [][]int{}
	var backTrack func(start int)
	backTrack = func(start int) {
		//  满足结束条件,到达树底
		if len(track) == k {
			res = append(res, append(make([]int, 0, 2), track...))
			return
		}
		for i := start; i < n+1; i++ {
			// 做选择
			track = append(track, i)
			// dfs
			backTrack(i + 1)
			// 撤销选择
			track = track[:len(track)-1]
		}
	}
	backTrack(1)
	return res
}

func main() {
	n := 4
	k := 2
	result := combine(n, k)
	fmt.Println(result)
}
